sympy/ntheory/factor_.py

```diff
@@ -502,29 +502,73 @@ def perfect_power(n, candidates=None, big=True, factor=True):
     sympy.core.intfunc.integer_nthroot
     sympy.ntheory.primetest.is_square
     """
-    if isinstance(n, Rational) and not n.is_Integer:
-        p, q = n.as_numer_denom()
-        if p is S.One:
-            pp = perfect_power(q)
-            if pp:
-                pp = (n.func(1, pp[0]), pp[1])
-        else:
-            pp = perfect_power(p)
-            if pp:
-                num, e = pp
-                pq = perfect_power(q, [e])
-                if pq:
-                    den, _ = pq
-                    pp = n.func(num, den), e
-        return pp
+    if isinstance(n, Rational) and not isinstance(n, Integer):
```

Comment:
Negative rational numbers are not taken into account. For example, `Rational(-1,2)**3`.

Reply:
In the version I gave, I had the negative handling first.

Reply:
Yes, I kept the original order of the blocks handling rationals and negative integers which makes the function incapable of handling negative rationals indeed. 

Keeping the order as mentioned in https://github.com/sympy/sympy/pull/27239#discussion_r1835812337 fixes that efficiently but what gets a bit problematic is, then the builtin integer casting of the argument via the line `n= as_int(n)` is missed and that causes a failure in the test `raises(ValueError, lambda: perfect_power(0.1))`. I hope I'm not wrong here in catching that. Please correct me if I am. I did try to overcome it in a naive way to check if I'm correct by putting the lines:
```python
if not isinstance(n, Rational):
     n = as_int(n)
```
at the very top of the function since I just thought that if the input ain't a rational number (`q != 1` though) then it shall be an integer so we should do the casting. And this did resolve the test failure.

Another test failure happens upon using that approach and I didn't quite get why that happens but it is logged as follows:

```
sympy/ntheory/tests/test_generate.py:64: in test_composite
    assert composite(4096) == 4736
sympy/ntheory/generate.py:1113: in composite
    b = int(n*(log(n) + log(log(n)))) # Upper bound for the search.
sympy/core/expr.py:308: in __int__
    r = self.round(2)
sympy/core/expr.py:3930: in round
    sign = 1 if x > 0 else -1
sympy/core/decorators.py:248: in _func
    return func(self, other)
sympy/core/expr.py:360: in __gt__
    return StrictGreaterThan(self, other)
sympy/core/relational.py:857: in __new__
    return cls._eval_relation(lhs, rhs, **options)
sympy/core/relational.py:864: in _eval_relation
    val = cls._eval_fuzzy_relation(lhs, rhs)
sympy/core/relational.py:1173: in _eval_fuzzy_relation
    return is_gt(lhs, rhs)
sympy/core/relational.py:1278: in is_gt
    return fuzzy_not(is_le(lhs, rhs, assumptions))
sympy/core/relational.py:1286: in is_le
    return is_ge(rhs, lhs, assumptions)
sympy/core/relational.py:1392: in is_ge
    n2 = _n2(lhs, rhs)
sympy/core/relational.py:1226: in _n2
    if a.is_comparable and b.is_comparable:
sympy/core/basic.py:898: in is_comparable
    for p in self.as_real_imag()]
sympy/core/add.py:998: in as_real_imag
    re, im = term.as_real_imag(deep=deep)
sympy/core/mul.py:855: in as_real_imag
    r, i = a.as_real_imag()
sympy/functions/elementary/exponential.py:893: in as_real_imag
    sarg = self.args[0].expand(deep, **hints)
sympy/core/cache.py:72: in wrapper
    retval = cfunc(*args, **kwargs)
sympy/core/expr.py:3678: in expand
    expr, hit = Expr._expand_hint(expr, hint, deep=deep, **hints)
sympy/core/expr.py:3617: in _expand_hint
    newexpr = getattr(expr, hint)(**hints)
sympy/functions/elementary/exponential.py:813: in _eval_expand_log
    p = perfect_power(arg)
sympy/ntheory/factor_.py:574: in perfect_power
    return _perfect_power(n)
sympy/ntheory/factor_.py:330: in _perfect_power
    n = _factorint_small(factors, n, 1_000, 1_000, next_p)[0]
sympy/ntheory/factor_.py:1018: in _factorint_small
    m = bit_scan1(n)
sympy/external/ntheory.py:28: in bit_scan1
    z = x.bit_length() - 1
E   AttributeError: 'Integer' object has no attribute 'bit_length'
```